### ISSUE TITLE
comment.author.role is blank

### ISSUE THREAD

**Issue by stvnrynlds** [timber/timber#1866]
### Expected behavior
Per documentation on https://timber.github.io/docs/reference/timber-comment/#author, using `{{comment.author.role}}` within a comment loop should print the applicable role for a commenter.

### Actual behavior
`{{comment.author.role}}` prints nothing.

#### twig input
```twig
<h3>Comments by...</h3>
<ol>
{% for comment in post.comments %}
    <li>{{comment.author.name}}, who is a {{comment.author.role}}</li>
{% endfor %}
</ol>
```
#### html result
```html
<h3>Comments by...</h3>
<ol>
    <li>Jared Novack, who is a </li>
    <li>Katie Ricci, who is a </li>
    <li>Rebecca Pearl, who is a </li>
</ol>
```

### What version of WordPress, PHP and Timber are you using?
Version 4.9.8, PHP 7.2.0, Timber 1.8.3

### How did you install Timber? (for example, from GitHub, Composer/Packagist, WP.org?)
Upgraded to newest version via plugin updater in WordPress dashboard.

**Comment by gchtr** [comment]
Clearly something’s missing 😅. Maybe it was never there.

@palmiak I think the relevant commits are these two:

- https://github.com/timber/timber/commit/14655bbb1e1f266184e6bb2ba6ed04906c152010
- https://github.com/timber/timber/commit/c2a0309fa7f23a615bd67723f9f1cbea567d4094

When we start thinking about how to make roles accessible in the Timber user object, I’d add it as a method. Simply accessing the roles array won’t work well, because this array contains "internal" slugs. They are not the real names of the roles. And they don’t work with internationalization. They will always be `administrator`, `editor` and `author`, and never `Administrator`, `Redakteur` or `Autor`, when we take a German example.

I’m thinking of this:

```php
/**
 * Gets the role of a user.
 *
 * @api
 * @since 1.8.4
 *
 * @return string|null
 */
public function role() {
    if ( empty( $this->roles ) ) {
        return null;
    }

    $wp_roles = wp_roles();
    $names    = $wp_roles->get_names();
    $role     = current( $this->roles );

    if ( isset( $names[ $role ] ) ) {
        return translate_user_role( $names[ $role ] );
    }

    return $role;
}
```

But then again, would somebody want to access the slug? Would a second method make sense? Maybe `role_slug()`? Or should `role()` have a parameter so that we can decide whether it returns a slug or not?

```php
$user->role(); // Editor
$user->role('slug'); // editor
```

In any way, when we import the `$roles` from the user object in `User::init()`, we should define the visibility of that property:

```php
/**
 * The roles the user is part of.
 *
 * @api
 * @since 1.8.4
 *
 * @var array
 */
protected $roles;
```

Here’s where I’m not entirely sure. Maybe it would be a good idea to set the visibility to `protected` instead of `public`, because especially with roles and capabilities in WordPress, it’s never a good idea to access these values directly. It’s always better to use the proper method. I see many developers dumping out an object and just using the values they find. They would probably do the same with the roles.

Speaking of capabilities. While we’re at it, we could also add a `can()` method:

```php
/**
 * Checks whether a user has a capability.
 *
 * @api
 * @since 1.8.4
 *
 * @param string $capability The capability to check.
 * @return bool Whether the user has the capability.
 */
public function can( $capability ) {
    return user_can( $this->ID, $capability );
}
```

**Comment by palmiak** [comment]
You're right about the translation - I didn't saw that :)

First of all - in theory one user can have many roles - so we should return an array - I was thinking about something like this:
```php
/**
 * @api
 * @var  array User roles
 */
protected $roles;

// in init()
if ( isset($data) && is_object($data) ) {
	if ( isset($data->data) ) {
		$this->import($data->data);
	} else {
		$this->import($data);
	}

	if ( isset($data->roles) ) {
		$this->roles = $this->get_roles($data->roles);
	}
}

/**
 * Create the array with user roles names and slugs.
 *
 * @api
 * @since 1.8.4
 *
 * @return array|null
 */
public function get_roles( $roles ) {
	if ( empty( $roles ) ) {
		return null;
	}

	$wp_roles = wp_roles();
	$names    = $wp_roles->get_names();

	$values = array();

	foreach( $roles as $role ) {
		$name = $role;
		if ( isset( $names[ $role ] ) ) {
			$name = translate_user_role( $names[ $role ] );
		}

		$values[] = array(
			'slug' => $role,
			'name' => $name,
		);
	}
	
	return $values;
}

/**
 * Gets the user roles.
 *
 * @api
 * @since 1.8.4
 *
 * @return array|null
 */
public function roles() {		
	return $this->roles;
}
```

with this user can get roles like this:
```twig
{% for userrole in comment.author.roles %}
    {{ userrole.name }}
    {{ userrole.slug }}
{% endfor %}
```

And the `can()` method is a great idea.

**Comment by stvnrynlds** [comment]
@palmiak @gchtr— Thanks for jumping on this!

For context, the reason I was reaching for `comment.author.role` was to identify whether or not a comment author was also a registered user on the site.

We have multiple accounts set up to respond to user comments on our blog, and I wanted to identify them with a class to show that they were official responses from the company, and not just guest comments.

If the plan is to return an array of roles, would a comment author who was not a registered user just return an empty array?

**Comment by gchtr** [comment]
@palmiak Sure, multiple roles! Of course we should also account for that! It’s always better when more people think about it 😁.

@stvnrynlds That’s a very common use case you have there. We should totally account for this, and probably also use it as an example somewhere!

When we have an array with `name` and `slug` keys, it’s hard to check for this. If we had an associative array, it could already work better:

```twig
{% if 'editor' in comment.author.roles|keys %}
```

Listing roles could work like this:

```twig
{% for slug, name in comment.author.roles %}
    {{ name }}
    {{ slug }}
{% endfor %}
```

I think this could also be used:

```twig
{% if comment.author.can('editor') %}
```

But, the documentation for [`WP_User::has_cap()`](https://developer.wordpress.org/reference/classes/wp_user/has_cap/) says:

> While checking against a role in place of a capability is supported in part, this practice is discouraged as it may produce unreliable results.

So it could also work with 

```twig
{% if comment.author.can('moderate_comments') %}
```

As long as you only do visual changes, like setting a certain CSS class for users from the moderation team, it’s probably fine to use `can('editor')`. But as soon as you want to display links to moderate comments in the frontend, `can('moderate_comments')` would have to be used.

So I’m thinking, do we actually need to expose the roles? Would there be a valid use case? 🤔

> If the plan is to return an array of roles, would a comment author who was not a registered user just return an empty array?

No, probably not because `{{ comment.author }}` always returns a User object. For users that are not registered on a site, the default role will be `subscriber`, as far as I know.

**Comment by palmiak** [comment]
> So I’m thinking, do we actually need to expose the roles? Would there be a valid use case? 🤔
Only showing user which role they are on some access limiting sites (paywalls for example). 

In other use cases `has_cap()` and some logic can do the job. 

> No, probably not because {{ comment.author }} always returns a User object. For users that are not 
> registered on a site, the default role will be subscriber, as far as I know.
As I remember it returned w User object with empty `roles` array inside. But that's from top of my head and I might be mistaken.

**Comment by palmiak** [comment]
@gchtr what do you think about this? For sure we should add `has_cap() / can()` - this one will be useful for sure.

About the rest - I see some limited use cases, but I agree it's not crucial. Although we written the whole code :)

**Comment by gchtr** [comment]
@palmiak I’d add all the functions, including a function to get a list of all roles. Because your paywall example makes total sense as well. For names, I’d go for

I think that one of the most crucial part here will to add a good amount of documentation and examples, including how to not use these functions. We already came up with good examples in this issue that we can use.

Do you want to make a pull request?

**Comment by palmiak** [comment]
No problem. We prepared all the code 😀

czw., 20 gru 2018, 19:47: Lukas Gächter <notifications@github.com>
napisał(a):

**Comment by jarednova** [comment]
Closed by #1898
